grafana/grafana
# Loki: Introduce `$__auto` range variable for metric queries

**What is this feature?**

We are introducing `$__auto` variable for Loki. This is similar for what we have done in Prometheus with `$__rate_interval`. 

**Why do we need this feature?**

In Loki, you need to set “step” usually at 2 places - in  [range] and in step. However, most of the time, you want [range] to be:
- Same as step for range queries
- Same as time range for instant queries

After discussion with the Loki team, we agreed that we would like to introduce $__auto range, which would default to values mentioned above. $__auto variable naming is consistent with Grafana variable syntax and it would be clear that is is variable. 

**Who is this feature for?**
For Loki users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/71743

**How to test this:**
1. Initialise Loki with `make devenv sources=loki`
2. Open Explore and run queries (both types - Instant and Range ) that include `[$__auto]` e.g. `rate({job="grafana"}[$__auto)` and ensure that queries are successfully executed.
3. Check in Code editor that $__auto is correctly validated - no red squiggly line and that it is offered as autocomplete suggestion
4. Check in Query builder that when you select `Rate` operation, it will give you $__auto as default value. Here also hover over "Explain this Operation" to see if $__auto has more detailed explanation. 
5. Check that in Kick start your query metric queries use $__auto and not $__interval
6. Check in label browser that when you click "Show logs rate", it uses $__auto

https://github.com/grafana/grafana/assets/30407135/ddc34b26-bb11-4dc3-baf8-1d2523d23816

**Special notes for your reviewer:**

We have decided to remove `$__interval` and `$__range` variables from the suggestions because `$__auto` replaces both and we don't want to confuse users with multiple options.

Please check that:
